pkg/reconciler/workload/workloadcluster/workloadcluster_controller.go

```diff
@@ -0,0 +1,153 @@
+/*
+Copyright 2022 The KCP Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package workloadcluster
+
+import (
+	"context"
+	"fmt"
+	"net/url"
+	"path"
+	"time"
+
+	"k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/klog/v2"
+
+	virtualworkspacesoptions "github.com/kcp-dev/kcp/cmd/virtual-workspaces/options"
+	workloadv1alpha1 "github.com/kcp-dev/kcp/pkg/apis/workload/v1alpha1"
+	kcp "github.com/kcp-dev/kcp/pkg/client/clientset/versioned"
+	"github.com/kcp-dev/kcp/pkg/client/informers/externalversions/workload/v1alpha1"
+	synceroptions "github.com/kcp-dev/kcp/pkg/virtual/syncer/options"
+)
+
+const controllerName = "kcp-workloadcluster-controller"
+
+func NewController(
+	kcpClient kcp.ClusterInterface,
+	workloadClusterInformer v1alpha1.WorkloadClusterInformer,
+	kcpServer *url.URL,
```

Comment:
I think this is not enough. We eventually need one URL per shard, i.e. add it to ClusterWorkspaceShard.status as a virtual workspace URL.